api/provider/aws/environment_scaler.go

```diff
@@ -1,14 +1,656 @@
+	"github.com/quintilesims/layer0/common/errors"
+	"github.com/quintilesims/layer0/common/models"
+)
+
+var defaultPorts = []int{
+	22,
+	2376,
+	2375,
+	51678,
+	51679,
+}
 
 type EnvironmentScaler struct {
+	Client              *awsc.Client
+	EnvironmentProvider provider.EnvironmentProvider
+	ServiceProvider     provider.ServiceProvider
+	TaskProvider        provider.TaskProvider
+	JobStore            job.Store
+	Config              config.APIConfig
+	deployCache         map[string][]*models.Resource
 }
 
-func NewEnvironmentScaler() *EnvironmentScaler {
-	return &EnvironmentScaler{}
+const (
+	Memory = iota
+	CPU
+)
+
+func NewEnvironmentScaler(a *awsc.Client, e provider.EnvironmentProvider, s provider.ServiceProvider, t provider.TaskProvider, j job.Store, c config.APIConfig) *EnvironmentScaler {
+	return &EnvironmentScaler{
+		Client:              a,
+		EnvironmentProvider: e,
+		ServiceProvider:     s,
+		TaskProvider:        t,
+		JobStore:            j,
+		Config:              c,
+	}
 }
 
 func (e *EnvironmentScaler) Scale(environmentID string) error {
-	return fmt.Errorf("EnvironmentScaler not implemented")
+	// GET RESOURCE PROVIDERS
+	clusterName := addLayer0Prefix(e.Config.Instance(), environmentID)
+
+	resourceProviders, err := e.getResourceProviders(clusterName)
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[DEBUG] resourceProviders for env '%s': %#v", environmentID, resourceProviders)
+
+	// GET RESOURCE CONSUMERS
+	resourceConsumers, err := e.getResourceConsumers(clusterName)
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[DEBUG] resourceConsumers for env '%s': %#v", environmentID, resourceConsumers)
+
+	_, err = e.scale(clusterName, resourceProviders, resourceConsumers)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (e *EnvironmentScaler) scale(clusterName string, providers []*models.Resource, consumers []*models.Resource) (*models.ScalerRunInfo, error) {
+	var errs []error
+
+	// determine most efficient method of packing tasks by cpu or memory.
+	e.determineSortPriority(providers, consumers, clusterName)
+
+	// check if we need to scale down
+	unusedProviders := []*models.Resource{}
+	for _, provider := range providers {
+		if !provider.InUse {
+			unusedProviders = append(unusedProviders, provider)
+		}
+	}
+
+	desiredScale := len(providers) - len(unusedProviders)
+
+	actualScale, err := e.scaleTo(clusterName, desiredScale, unusedProviders)
+	if err != nil {
+		errs = append(errs, err)
+	}
+
+	scaleBeforeRun := len(providers)
+	info := &models.ScalerRunInfo{
+		EnvironmentID:           clusterName,
+		PendingResources:        resourceModels(consumers),
+		ResourceProviders:       resourceModels(providers),
+		ScaleBeforeRun:          scaleBeforeRun,
+		DesiredScaleAfterRun:    desiredScale,
+		ActualScaleAfterRun:     actualScale,
+		UnusedResourceProviders: len(unusedProviders),
+	}
+
+	return info, errors.MultiError(errs)
+}
+
+func resourceModels(resources []*models.Resource) []models.Resource {
```

Comment:
This function needs a better name

Reply:
why does this function exist?

Reply:
We don't know - this was something we found in `develop` and we weren't sure what the heck the point of it was.

I've removed it over in #440.

✅